[WIP] Adding agent provider docs to hosted control planes docs (#4628)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/configure_hosted.adoc

[#hosted-control-planes-configure]
= Configuring hosted control planes (Technology Preview)

To configure hosted control planes, you need a hosting cluster and a hosted cluster. By deploying the HyperShift Operator on a managed cluster by using the `hypershift-addon` managed cluster add-on, you can make that cluster into a hosting cluster and start to create a hosted cluster. The {mce} 2.2 supports only the default `local-cluster` and the hub cluster as the hosting cluster.

> Hosted control planes is a Technology Preview feature, so the related components are disabled by default. See the following topics to learn more about how to enable hosted control planes on Amazon Web Services (AWS) or bare metal:

Reviewer: @zanetworker, Do you want to "market" the agent platform as bare metal or on-prem?

Reviewer 2: Baremetal as there is no intention to use the agent for other providers today. If there need be a change in the future we can easily do that. If a user want's baremetal, I want them to immediately see how they can get there without having to decipher "agent" (for now).

Reviewer 3: seems this convo is resolved

Reviewer: Indeed

hypershift create cluster aws --help
----

[#hosting-service-cluster-configure-bm]
== Configuring the hosting cluster on bare metal

> To provision hosted control planes on bare metal, you can use the Agent platform. The Agent platform uses the Infrastructure Operator, which is also known as the Assisted Installer, to add worker nodes to a hosted cluster. For an introduction to the Infrastructure Operator, see https://github.com/openshift/assisted-service/blob/master/docs/hive-integration/kube-api-getting-started.md[Kube API - Getting Started Guide]. Each bare metal host must be started with a Discovery Image that the Infrastructure Operator provides. You can start the hosts manually or through automation by using the Cluster-Baremetal-Operator. After each host starts, it runs an Agent process to discover the host details and its installation. Each host is represented by an Agent custom resource.

Reviewer: ```suggestion
To provision hosted control planes on bare metal, you can use the Agent platform. The Agent platform uses the Central Infrastructure Management service (CIM) to add worker nodes to a hosted cluster. For an introduction to CIM, see https://github.com/stolostron/rhacm-docs/blob/2.7_stage/clusters/cluster_lifecycle/create_infra_env.adoc[here]. Each bare metal host must be started with a Discovery Image that the Infrastructure Operator provides. You can boot the hosts manually or through automation by using the Cluster-Baremetal-Operator. After each host starts, it runs an Agent process to discover the host details and perform its installation. An Agent custom resource represents each host.
```

Reviewer 4: Made some of these updates manually. I tried to make the tie here between AI and CIM. Would they still be using CIM with mce and not ACM? 

Clarified this during the meeting. Thanks.

When you create a hosted cluster with the Agent platform, HyperShift installs the Agent CAPI provider in the Hosted Control Plane (HCP) namespace.

When you scale up a node pool, a machine is created, and the CAPI provider finds an Agent that is approved, is passing validations, is not currently in use, and meets the requirements that are specified in the node pool spec. You can monitor the installation of an Agent by checking its status and conditions.

> When you scale down a node pool, Agents are unbound from the corresponding cluster. However, before you re-use them, you must boot them with the Discovery Image again.

Reviewer: If the host is managed by a BareMetalHost (Cluster-Baremetal-Operator), that automation will be used to reboot the host.
Also, we recently implemented a feature where if the host was successfully installed, it should be able to boot the Discovery Image itself automatically after being unbound.  I will post the version where this functionality was added.

Reviewer: This non-BMH feature is new in ACM 2.7
https://issues.redhat.com/browse/MGMT-7503

Reviewer: In short, I think we can change this to:

```suggestion
When you scale down a node pool, Agents are unbound from the corresponding cluster. CIM will automatically boot them with the Discovery Image so they may be used again.
```

Reviewer 5: Wait, keep the current text, we just found this
https://issues.redhat.com/browse/MGMT-13502

Reviewer 4: Edited the docs, but tried to leave the original intent.

[#hosting-service-cluster-configure-prereq-bm]
=== Prerequisites

You must have the following prerequisites to configure a hosting cluster: 

* {mce} 2.2 and later installed on an {ocp-short} cluster. When you install {product-title-short}, {mce-short} is automatically installed. You can also install {mce-short} without {product-title-short} as an Operator from the {ocp-short} OperatorHub.

* The hosted control planes command line interface (CLI) as a plugin for `oc` to create and manage the hosted cluster in {mce}. To install the CLI, follow the instructions in <<installing-the-hosted-control-plane-cli,Installing the hosted control plane CLI>>.

* {mce} must have at least one managed {ocp-short} cluster. The `local-cluster` is automatically imported in {mce} 2.2 and later. See xref:../install_upgrade/adv_config_install.adoc#advanced-config-engine[Advanced configuration] for more information about the `local-cluster`. You can check the status of your hub cluster by running the following command:
+
----
oc get managedclusters local-cluster
----

[#install-assisted-service-and-hive]
=== Installing the Assisted Service and Hive Operators

> If {product-title} is installed, you can skip this step because the Infrastructure Operator and Hive Operator are included in that product.

Reviewer: ```suggestion
If {product-title} is installed, you can skip this step because the Central Infrastructure Management service and Hive Operator are included in that product.
```

Reviewer 4: Manually updated.

Reviewer 2: We don't need Hive Operators do we @avishayt ?

Reviewer 3: Waiting to see an update about this.

Reviewer: No, text is fine now.

To deploy the Operators, you can use https://github.com/karmab/tasty[tasty], which is a CLI that handles Operators.

. Install tasty by entering the following command:
+
----
curl -s -L https://github.com/karmab/tasty/releases/download/v0.4.0/tasty-linux-amd64 > ./tasty
sudo install -m 0755 -o root -g root ./tasty /usr/local/bin/tasty
----

. Install the Operators by entering the following command:
+
----
tasty install assisted-service-operator hive-operator
----

[#configure-agent-service]

> === Configuring the Agent Service

Reviewer: ```suggestion
=== Enabling the Central Infrastructure Management service
```

Link to here - https://github.com/stolostron/rhacm-docs/blob/2.7_stage/clusters/cluster_lifecycle/create_infra_env.adoc#enabling-the-central-infrastructure-management-service

Reviewer 3: This section also should not be a headline with a link... If this is a next step, we need to write it as such.   

Enable the Central Infrastructure Management service. See (link) for instructions...

That is how we would write it. The extra header is not needed. Does that work? @avishayt

Reviewer 3: I changed it to fit the format of a step, like all other steps.

Reviewer 5: Looks good, it should be just a step under `Installing the Assisted Service and Hive Operators`
I think the text should be just like other steps and shouldn't start "You need to"
e.g.
```
 . Enable the Central Infrastructure Management service. See xref:../cluster_lifecycle/create_infra_env.adoc#enble-cim[Enabling the Central Infrastructure Management service] for the procedure.
```

Create the `AgentServiceConfig` resource as follows:
----
export DB_VOLUME_SIZE="10Gi"
export FS_VOLUME_SIZE="10Gi"
export OCP_VERSION="4.11.5"
export OCP_MAJMIN=${OCP_VERSION%.*}

> export ARCH="x86_64"

Reviewer 3: 325 and below should not be bold in pr source, the result of code on line 324 I have been trying to keep clean in our repo.

When you create a hosted cluster with the Agent platform, HyperShift installs the Agent CAPI provider in the Hosted Control Plane (HCP) namespace.

When you scale up a node pool, a machine is created, and the CAPI provider finds an Agent that is approved, is passing validations, is not currently in use, and meets the requirements that are specified in the node pool spec. You can monitor the installation of an Agent by checking its status and conditions.

When you scale down a node pool, Agents are unbound from the corresponding cluster. However, before you re-use them, you must boot them with the Discovery Image again.

[#hosting-service-cluster-configure-prereq-bm]
=== Prerequisites

You must have the following prerequisites to configure a hosting cluster: 

* The {mce} 2.2 and later installed on an {ocp-short} cluster. The {mce} is automatically installed when you install {product-title-short}. The {mce} can also be installed without {product-title-short} as an operator from the {ocp-short} OperatorHub.

> * The hosted control planes command line interface as a plugin for `oc` to create and manage the hosted cluster in {mce}. Use one of the following methods to install the command line interface:

Reviewer 6: https://github.com/stolostron/rhacm-docs/blob/2.7_stage/clusters/hosted_control_planes/configure_hosted.adoc#installing-the-hosted-control-planes-cli

This is documented here.

** Navigate to your {ocp-short} console command line tools page. Select the *Hosted Control Plane CLI* tool and follow the instructions to set up the plugin.
** See https://github.com/stolostron/hypershift-addon-operator/blob/main/docs/installing_hypershift_cli.md[Getting started with Hypershift Hosted Control Plane CLI] and follow the instructions.
** Clone and build the binary file from the https://github.com/openshift/hypershift[HyperShift repository].

* {mce} must have at least one managed {ocp-short} cluster. The `local-cluster` is automatically imported in {mce} 2.2 and later. See xref:../install_upgrade/adv_config_install.adoc#advanced-config-engine[Advanced configuration] for more information about the `local-cluster`. You can check the status of your hub cluster by running the following command:
+
----
oc get managedclusters local-cluster
----

> [#install-hypershift-operator]

Reviewer 6: This should point to https://github.com/stolostron/rhacm-docs/blob/2.7_stage/clusters/hosted_control_planes/configure_hosted.adoc#installing-the-hosted-control-planes-cli

=== Install the HyperShift Operator

> . Install the HyperShift CLI either by building it or extracting it from the Operator image.

Reviewer 2: Looks like this overlaps with the above step of using the CLI as an `oc` plugin. 

Also, this does not guarantee that cli plugin will work because that would pull from latest.

Author: Good catch. Should I remove this step (lines 302 - 322) since the prereqs already cover how to get the CLI?

* To build the CLI, enter the following commands by using Go 1.18:
+
----   
shell linenums="1"
git clone https://github.com/openshift/hypershift.git
cd hypershift
make build
sudo install -m 0755 bin/hypershift /usr/local/bin/hypershift
----

* To extract the CLI from an Operator image, enter commands that are like the commands in the following example. This example uses Podman.       
+
----
export HYPERSHIFT_RELEASE=4.11
  
podman cp $(podman create --name hypershift --rm --pull always quay.io/hypershift/hypershift-operator:${HYPERSHIFT_RELEASE}):/usr/bin/hypershift /tmp/hypershift && podman rm -f hypershift
  
sudo install -m 0755 -o root -g root /tmp/hypershift /usr/local/bin/hypershift
----

. Deploy the HyperShift Operator.
+

> **Important:** If you do not define the HyperShift image that you want to use, by default, the CLI deploys the latest image. Typically, you want to deploy the image that matches the release of the {ocp-short} cluster where you plan to run HyperShift.

Reviewer 2: We should not leave the image up for choice, that's a main difference from upstream docs.

Author: Good to know. I can update the text here to instruct users to deploy the image that matches the release of OCP where they want to run HyperShift.

Reviewer 6: https://github.com/stolostron/rhacm-docs/blob/2.7_stage/clusters/hosted_control_planes/configure_hosted.adoc#installing-the-hosted-control-planes-cli

The CLI binary should be installed by following the doc above.

** To install the latest image, enter the following command:
+
----
hypershift install
----

** To deploy the image that matches the release of the cluster where you plan to run HyperShift, enter the following command:
+
----

> hypershift install --hypershift-image quay.io/hypershift/hypershift-operator:4.11

Reviewer 2: why 4.11?

Author: I copied that from the upstream docs. I can change it to 4.12 or replace it with a variable and indicate that the user should replace the variable with the release where they want to run HyperShift.

Reviewer 2: The user should be given a validated release vs. choose. Added a thread on this on slack to follow up :)

----

. Verify that you see the operator running in the `hypershift` namespace:
+
----
oc -n hypershift get pods

NAME                      READY   STATUS    RESTARTS   AGE
operator-55fffbd6-whkxs   1/1     Running   0          61s
----

[#install-assisted-service-and-hive]
=== Install the Assisted Service and Hive Operators

If {product-title} is installed, you can skip this step because the Infrastructure Operator and Hive Operator are included in that product.

> To deploy the Operators, you can use https://github.com/karmab/tasty[tasty].

Reviewer 2: What is this "https://github.com/karmab/tasty[tasty]" link?

Author: it's a link to tasty. That text will render as "To deploy the Operators, you can use [tasty](https://github.com/karmab/tasty)."

Reviewer 2: The link above 404s for me.

Author: hmm..it works for me. Does this open a 404 for you? https://github.com/karmab/tasty

Reviewer 2: The one in the doc yes: https://github.com/karmab/tasty[tasty]

> To deploy the Operators, you can use https://github.com/karmab/tasty[tasty].

Reviewer 2: ```suggestion
To deploy the Operators, you can use https://github.com/karmab/tasty.
```

Author: If I apply this suggested change, the output won't render an inline link. The text in brackets is necessary. It's just really hard to see that in the GitHub interface.

[#configure-dns-bm]
=== Configuring DNS

The API Server for the hosted cluster is exposed a `NodePort` service. A DNS entry must exist for `api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN}` that points to destination where the API Server can be reached.

The DNS entry can be as simple as a record that points to one of the nodes in the management cluster that is running the hosted control plane. The entry can also point to a load balancer that is deployed to redirect incoming traffic to the ingress pods. See link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.12/html/installing/installing-on-any-platform#installation-load-balancing-user-infra_installing-platform-agnostic[Load balancing requirements for user-provisioned infrastructure] for more information about load balancers.

.Example DNS configuration
----
api.example.krnl.es.    IN A 192.168.122.20
api.example.krnl.es.    IN A 192.168.122.21
api.example.krnl.es.    IN A 192.168.122.22
api-int.example.krnl.es.    IN A 192.168.122.20
api-int.example.krnl.es.    IN A 192.168.122.21
api-int.example.krnl.es.    IN A 192.168.122.22
*.apps.example.krnl.es. IN A 192.168.122.23
----

[#deploying-sr-iov]

> == Deploying the SR-IOV Operator

Reviewer 4: @zanetworker @avishayt @rokej The way this section is set up, it looks like you have to deploy the SR-IOV Operator on either AWS or bm. Is that correct?

Reviewer 6: The referenced doc says I need to configure the SR-IOV operator for any hosted cluster I create. I am asking the hypershift channel about this requirement https://redhat-internal.slack.com/archives/C01C8502FMM/p1675353294985239

Reviewer 6: @cadawson @zanetworker confirmed that it is an optional setup for a hosted cluster. We should just remove the whole `Deploying the SR-IOV Operator` section.

Reviewer 2: This does not need to be mentioned here. OCP features related to OCP operators live on docs.ocp 

For these docs, we want to make sure we cover how to lifecycle clusters (for BM that is create/destroy)

Reviewer 3: We should not have a headline/section that just points to a link. Editing this

Reviewer 6: Remove this `Deploying the SR-IOV Operator` section.

Reviewer 3: I did remove it, but kept the link.

Reviewer 3: @rokej are you saying we don't need any of it?

Reviewer 6: @swopebe I saw the whole updated doc and it looks fine as-is now. Please ignore my previous comment. Thanks.

[#hosting-service-cluster-configure-prereq-bm]
=== Prerequisites

You must have the following prerequisites to configure a hosting cluster: 

* {mce} 2.2 and later installed on an {ocp-short} cluster. When you install {product-title-short}, {mce-short} is automatically installed. You can also install {mce-short} without {product-title-short} as an Operator from the {ocp-short} OperatorHub.

* The hosted control planes command line interface (CLI) as a plugin for `oc` to create and manage the hosted cluster in {mce}. To install the CLI, follow the instructions in xref:../hosted_control_planes/configure_hosted.adoc#hosted-install-cli[Installing the hosted control plane CLI.

* {mce} must have at least one managed {ocp-short} cluster. The `local-cluster` is automatically imported in {mce} 2.2 and later. See xref:../install_upgrade/adv_config_install.adoc#advanced-config-engine[Advanced configuration] for more information about the `local-cluster`. You can check the status of your hub cluster by running the following command:
+
----
oc get managedclusters local-cluster
----

[#install-assisted-service-and-hive]

> === Installing the Assisted Service and Hive Operators
> 
> If {product-title} is installed, you can skip this step because the Central Infrastructure Management service and Hive Operator are included in that product.
> 
> To deploy the Operators, you can use https://github.com/karmab/tasty[tasty], which is a CLI that handles Operators.
> 
> . Install tasty by entering the following command:
> +
> ----
> curl -s -L https://github.com/karmab/tasty/releases/download/v0.4.0/tasty-linux-amd64 > ./tasty
> sudo install -m 0755 -o root -g root ./tasty /usr/local/bin/tasty
> ----
> 
> . Install the Operators by entering the following command:
> +
> ----
> tasty install assisted-service-operator hive-operator
> ----
> 
> . You need to enable the Central Infrastructure Management service to create hosted clusters in a bare metal environment. See xref:../cluster_lifecycle/create_infra_env.adoc#enble-cim[Enabling the Central Infrastructure Management service] for the procedure. 

Reviewer: ```suggestion
=== Enabling the Central Infrastructure Management service

The Central Infrastructure Management service is included as part of {product-title}.  To enable the service see xref:../cluster_lifecycle/create_infra_env.adoc#enble-cim[Enabling the Central Infrastructure Management service] for the procedure. 
```
